backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/user/UserService.java

```diff
@@ -153,4 +168,149 @@ public void resetDemoStudents() {
                     this.userRepository.delete(user);
                 });
     }
+
+    @Retryable(
+            value = { SQLException.class },
+            backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public void importListOfUsers(InputStream stream, int courseExecutionId) {
+        String line = "";
+        String cvsSplitBy = ",";
+        User.Role auxRole;
+        InputStreamReader isr = new InputStreamReader(stream, StandardCharsets.UTF_8);
+        try (BufferedReader br = new BufferedReader(isr)) {
+            while ((line = br.readLine()) != null) {
+                String[] userInfo = line.split(cvsSplitBy);
+                if (userInfo.length == 2) {
+                    auxRole = User.Role.STUDENT;
+                }
+                else if (userInfo.length > 2) {
+                    if (userInfo[2].equals("STUDENT")) {
+                        auxRole = User.Role.STUDENT;
+                    }
+                    else {
+                        auxRole = User.Role.TEACHER;
+                    }
+                }
+                else {
+                    throw new TutorException(INVALID_CSV_FILE_FORMAT);
+                }
+                ExternalUserDto userDto = new ExternalUserDto();
+                userDto.setEmail(userInfo[0]);
+                userDto.setName(userInfo[1]);
+                userDto.setRole(auxRole);
+                createExternalUser(courseExecutionId, userDto);
+            }
+        } catch (IOException ex) {
+            throw new TutorException(ErrorMessage.CANNOT_OPEN_FILE);
+        }
+    }
+
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public ExternalUserDto createExternalUser(Integer courseExecutionId, ExternalUserDto externalUserDto) {
+        verifyEmail(externalUserDto);
+        verifyRole(externalUserDto);
+        CourseExecution courseExecution = getCourseExecution(courseExecutionId);
+        User user1 = getUser(externalUserDto, courseExecution);
+        associateUserWithExecution(courseExecution, user1);
+        sendConfirmationEmailTo(user1);
```

Comment:
We are sending confirmation emails even if the transaction fails, when it is invoked from importListOfUsers. Only send the confirmation email when the transaction cannot abort. Decouple send email from user creation. Send the ones that are inactive from the list, note that some may already exist and are active.